src/hipscat/inspection/almanac.py

```diff
@@ -0,0 +1,255 @@
+            self._add_files_to_namespace(dirs)
+        elif dirs is not None:
+            self._add_files_to_namespace(dirs)
+
+    def _add_files_to_namespace(self, directory, namespace=""):
+        """Get almanac files within a directory or list of directories.
+
+        Args:
+            directory: directory to scan
+            namespace: if provided, files in this directory will be in their
+                own namespace in the almanac
+        """
+        if not pd.api.types.is_list_like(directory):
+            directory = [directory]
+
+        files = []
+        for input_path in directory:
+            if os.path.isfile(input_path):
+                files.append(input_path)
+                continue
+
+            input_paths = glob.glob(f"{input_path}/**.yml")
+            input_paths.sort()
+            files.extend(input_paths)
+
+        if namespace in self.files:
+            self.files[namespace].extend(files)
+        else:
+            self.files[namespace] = files
+
+    def _init_catalog_objects(self):
+        """Create (unlinked) almanac info objects for all the files found
+        in the previous steps."""
+        for namespace, files in self.files.items():
+            for file in files:
+                catalog_info = AlmanacInfo.from_file(file)
+                catalog_info.namespace = namespace
+                if namespace:
+                    full_name = f"{namespace}:{catalog_info.catalog_name}"
+                else:
+                    full_name = catalog_info.catalog_name
+                if full_name in self.entries:
+                    raise ValueError(
+                        f"Duplicate catalog name ({full_name}). Try using namespaces."
+                    )
+                self.entries[full_name] = catalog_info
+                self.dir_to_catalog_name[catalog_info.catalog_path] = full_name
+
+    def _init_catalog_links(self):
+        """Initialize the links between almanac catalogs.
+
+        For each type of link (e.g. primary or join), look for the catalog in
+        the almanac, using whatever text we have. If found, add the object
+        to the almanac info as a pointer. Additionally, add the reference to
+        the linked object, so catalogs know about each other from either side.
+        """
+        for catalog_entry in self.entries.values():
+            if catalog_entry.catalog_type == CatalogType.OBJECT:
+                ## Object currently has no links that start in the catalog.
+                pass
+            elif catalog_entry.catalog_type == CatalogType.SOURCE:
+                ## Source catalogs MAY indicate their primary object catalog.
+                if catalog_entry.primary:
+                    object_catalog = self._get_linked_catalog(
+                        catalog_entry.primary, catalog_entry.namespace
+                    )
+                    if not object_catalog:
+                        raise ValueError(
```

Comment:
Scattered thoughts below:

For the general case of someone who wants to explore the datasets and understand what data is available to them, I think we should enforce that almanac data is well-formatted, but it might be mean to crap out entirely on them because of one bad catalog.

For the case you mention, if someone wants to work on a specific, known catalog (be it object or source), then they don't really need to use the almanac to get at it. They can just use the hipscat/lsdb API directly with the catalog_base_dir. If they don't care about the affiliated tables, then the almanac would likely just get in their way.

But I've made these calls print warnings if linked catalogs aren't found, or the data is just inconsistent in some way. This should help users still make use of the almanac, and give them a chance to fix their data.

Reply:
totally makes sense